segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,378 @@
+        }
+
+        try {
+            if (Files.size(path) < offset) {
+                log.warn("Read called on segment {} at offset {}. The offset is beyond the current size of the file.",
+                        handle.getSegmentName(), offset);
+                throw new CompletionException(new ArrayIndexOutOfBoundsException());
+            }
+        } catch (IOException ioe) {
+            throw new CompletionException(ioe);
+        }
+
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) {
+            int bytesRead = channel.read(ByteBuffer.wrap(buffer, bufferOffset, length), offset);
+            return bytesRead;
+        } catch (IOException | IndexOutOfBoundsException e) {
+            if (e instanceof IndexOutOfBoundsException) {
+                throw new CompletionException(new ArrayIndexOutOfBoundsException(e.getMessage()));
+            } else {
+                throw new CompletionException(e);
+            }
+        }
+    }
+
+    private SegmentProperties syncGetStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        try {
+            PosixFileAttributes attrs = Files.readAttributes(Paths.get(config.getNfsRoot(), streamSegmentName),
+                    PosixFileAttributes.class);
+            StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName, attrs.size(),
+                    !(attrs.permissions().contains(OWNER_WRITE)), false,
+                    new ImmutableDate(attrs.creationTime().toMillis()));
+
+            return information;
+        } catch (IOException ioe) {
+            throw new CompletionException(ioe);
+        }
+    }
+
+    private boolean syncExists(String streamSegmentName, Duration timeout) {
+        boolean exists = Files.exists(Paths.get(config.getNfsRoot(), streamSegmentName));
+        return exists;
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName, Duration timeout) {
+        log.info("Creating Segment {}", streamSegmentName);
+        try {
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as rw-r--r-- 644
+            perms.add(PosixFilePermission.OWNER_WRITE);
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+            Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+            Files.createDirectories(path.getParent());
+            Files.createFile(path, fileAttributes);
+            log.info("Created Segment {}", streamSegmentName);
+            return this.getStreamSegmentInfo(streamSegmentName, timeout).get();
+        } catch (IOException | InterruptedException | ExecutionException e) {
```

Comment:
You need to handle the case when two overlapping requests to create segments of the same stream end up stepping on each other's toes. One of them may create the directories and the second one will fail because the directories are already created.